security/pkg/nodeagent/caclient/providers/google-cas/client.go

```diff
@@ -0,0 +1,157 @@
+	var err error
+	caClient := &GoogleCASClient{caSigner: capool, provider: provider}
+	ctx := context.Background()
+
+	if lis == nil {
+		caClient.caClient, err = privateca.NewCertificateAuthorityClient(ctx,
+			option.WithGRPCDialOption(grpc.WithPerRPCCredentials(provider)))
+	} else {
+		caClient.caClient, err = privateca.NewCertificateAuthorityClient(ctx,
+			option.WithoutAuthentication(),
+			option.WithGRPCDialOption(grpc.WithContextDialer(mock.ContextDialerCreate(lis))),
+			option.WithGRPCDialOption(grpc.WithInsecure()))
+	}
+	if err != nil {
+		googleCAClientLog.Errorf("unable to initialize google cas caclient: %v", err)
+		return nil, err
+	}
+	return caClient, nil
+}
+
+func (r *GoogleCASClient) createCertReq(name string, csrPEM []byte, lifetime time.Duration) *privatecapb.CreateCertificateRequest {
+	subjectConfig := &privatecapb.CertificateConfig_SubjectConfig{
+		// Empty Subject
+		Subject: &privatecapb.Subject{},
+	}
+	keyUsage := &privatecapb.KeyUsage{
+		BaseKeyUsage: &privatecapb.KeyUsage_KeyUsageOptions{
+			DigitalSignature: true,
+			KeyEncipherment:  true,
+		},
+		ExtendedKeyUsage: &privatecapb.KeyUsage_ExtendedKeyUsageOptions{
+			ServerAuth: true,
+			ClientAuth: true,
+		},
+	}
+
+	// Derive Public Key from the csrPEM
+	publicKey := &privatecapb.PublicKey{
+		// TODO: Need to get key-type from security options
+		Format: privatecapb.PublicKey_PEM,
+		Key:    csrPEM,
+	}
+
+	creq := &privatecapb.CreateCertificateRequest{
+		Parent:        r.caSigner,
+		CertificateId: name,
+		Certificate: &privatecapb.Certificate{
+			Lifetime: durationpb.New(lifetime),
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					SubjectConfig: subjectConfig,
+					X509Config: &privatecapb.X509Parameters{
+						KeyUsage: keyUsage,
+					},
+					PublicKey: publicKey,
+				},
+			},
+			SubjectMode: privatecapb.SubjectRequestMode_REFLECTED_SPIFFE,
+		},
+	}
+	return creq
+}
+
+// CSR Sign calls Google CAS to sign a CSR.
+func (r *GoogleCASClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) ([]string, error) {
+	certChain := []string{}
+	name := fmt.Sprintf("csr-workload-%s", rand.String(8))
+	creq := r.createCertReq(name, csrPEM, time.Duration(certValidTTLInSec)*time.Second)
+
+	ctx := context.Background()
+
+	cresp, err := r.caClient.CreateCertificate(ctx, creq)
+	if err != nil {
+		googleCAClientLog.Errorf("unable to create certificate: %v", err)
+		return []string{}, err
+	}
+	certChain = append(certChain, cresp.GetPemCertificate())
+	certChain = append(certChain, cresp.GetPemCertificateChain()...)
+	return certChain, nil
+}
+
+func (r *GoogleCASClient) GetRootCertBundle() ([]string, error) {
+	var rootCertMap map[string]struct{} = make(map[string]struct{})
+	var trustbundle []string = []string{}
+	var err error
+
+	ctx := context.Background()
+
+	req := &privatecapb.FetchCaCertsRequest{
+		CaPool: r.caSigner,
+	}
+	resp, err := r.caClient.FetchCaCerts(ctx, req)
```

Comment:
will this be called from individual workload sidecars? IIRC, individual workloads won't have permission to call this API, but only the control-plane identity (e.g. istiod)

Reply:
Yes, the intent is for this code to be directly invoked from all the individual workload sidecars. This would require some sort of Google workload identity to be enabled on the cluster + all workloads in the pool would need to have the IAM permissions to retrieve certificates/ca-certificates from a particular pool